File: Programmers/level3/단어 변환.cc
#include <string>
#include <vector>
#include <climits>

using namespace std;
string targetWord;
vector<string> words;
int answer = INT_MAX;

bool visited[52];

bool isOnlyOneDiff(string w1, string w2) {
    int diffCount = 0;
    for (int i = 0; i < (int)w1.size(); ++i) {
        if (w1[i] != w2[i]) diffCount++;
    }

    if (diffCount == 1) return true;
    else return false;
}

void dfs(string curWord, int curCount) {
    if (curWord == targetWord) {
        answer = min(answer, curCount);
        return;
    }

    for (int i = 0; i < (int)words.size(); ++i) {
        string nextWord = words[i];
        if (visited[i]) continue;
        if (isOnlyOneDiff(curWord, nextWord)) {
            visited[i] = true;
            dfs(nextWord, curCount + 1);
            visited[i] = false;
        }
    }
}

int solution(string b, string t, vector<string> ws) {
    targetWord = t;
    words = ws;

    dfs(b, 0);
    if (answer == INT_MAX) answer = 0;
    return answer;
}

File: Programmers/level3/단어변환.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

class Solution {
    int answer = Integer.MAX_VALUE;
    List<String> words = new ArrayList<>();
    boolean[] isUsed = new boolean[52];

    public int solution(String begin, String target, String[] wordss) {
        words = Arrays.stream(wordss).collect(Collectors.toList());

        dfs(begin, target, 0);

        if (answer == Integer.MAX_VALUE) return 0;
        return answer;
    }

    private void dfs(final String curWord, final String target, final int count) {
        if (curWord.equals(target)) {
            answer = Math.min(answer, count);
            return;
        }

        for (int i = 0; i < words.size(); ++i) {
            if (isUsed[i]) continue;
            if (isOnlyOneDiff(curWord, words.get(i))) {
                isUsed[i] = true;
                dfs(words.get(i), target, count + 1);
                isUsed[i] = false;
            }
        }
    }

    private boolean isOnlyOneDiff(final String curWord, final String nextWord) {
        int diffCount = 0;
        for (int i = 0; i < curWord.length(); ++i) {
            if (curWord.charAt(i) != nextWord.charAt(i)) diffCount++;
        }
        return diffCount == 1;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        String[] inputWords = {
                "hot", "dot", "dog", "lot", "log", "cog"
        };
        System.out.println(solution.solution("hit", "cog", inputWords));
    }
}
